src/distr/integer.rs

```diff
@@ -107,21 +107,50 @@ impl_nzint!(NonZeroI64, NonZeroI64::new);
 impl_nzint!(NonZeroI128, NonZeroI128::new);
 
 #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
-macro_rules! x86_intrinsic_impl {
-    ($meta:meta, $($intrinsic:ident),+) => {$(
-        #[cfg($meta)]
-        impl Distribution<$intrinsic> for StandardUniform {
-            #[inline]
-            fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> $intrinsic {
-                // On proper hardware, this should compile to SIMD instructions
-                // Verified on x86 Haswell with __m128i, __m256i
-                let mut buf = [0_u8; core::mem::size_of::<$intrinsic>()];
-                rng.fill_bytes(&mut buf);
-                // x86 is little endian so no need for conversion
-                zerocopy::transmute!(buf)
-            }
-        }
-    )+};
+impl Distribution<__m128i> for StandardUniform {
+    #[inline]
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> __m128i {
+        // NOTE: It's tempting to use the u128 impl here, but confusingly this
+        // results in different code (return via rdx, r10 instead of rax, rdx
+        // with u128 impl) and is much slower (+130 time). This version calls
+        // impls::fill_bytes_via_next but performs well.
+
+        let mut buf = [0_u8; core::mem::size_of::<__m128i>()];
+        rng.fill_bytes(&mut buf);
+        // x86 is little endian so no need for conversion
+
+        // SAFETY: both source and result types are valid for all values.
```

Comment:
Two things to note:
- This should say that any bit-valid `[u8; N]` is also a bit-valid `__m128i` ("all values" technically doesn't preclude uninitialized bytes, which are unsound to transmute to a `__m128i`); it should also say that they're the same size (while `mem::transmute` wouldn't permit this to compile, transmuting a value of a smaller type to one of a larger type would implicitly leave the trailing bytes uninitialized)
-  Being pedantic, this should provide a citation for the claim that any byte values are valid for `__m128i`; feel free to rip off our safety comment for this: https://github.com/google/zerocopy/blob/dccbbcd8714eebeeac794ca5dd1ba1cbecc09ea2/src/impls.rs#L815-L879

Reply:
If we want to be super pedantic and use MiniRust terminology, "value" refers to high-level concepts like mathematical integers, not not raw bytes stored in memory. Being "valid" is a property of byte sequences; invalid byte sequences do not even correspond to a value. The key for transmute safety is that all representations (i.e., byte sequences representing a value) of the source are also representations of the target.

That's how I would phrase it, anyway. We haven't officially adopted this terminology I think.